Docs: Memory management and writing tests (#4)
nanjekyejoannah/micropython: docs/develop/core.rst

The remaining arguments take on the same meaning.

.. note::

>    The function ``DEF_RULE`` or ``DEF_RULE_NC`` takes other arguments. For an indepth understanding

Reviewer: ```suggestion
   The function ``DEF_RULE`` or ``DEF_RULE_NC`` takes other arguments. For an in-depth understanding
```

   }

The difference here is we have to handle *viper* typing.

Memory Management
-----------------

> Contrary to programming languages like C/C++, MicroPython hides memory management details 

Reviewer: ```suggestion
Unlike many programming languages such as C/C++, MicroPython hides memory management details 
```

> from the developer by supporting automatic memory management (AMM).

Reviewer: Is 'AMM' a common acronym? I hadn't used it before...

AMM is a technique used by operating systems or applications to automatically manage the allocation and deallocation of memory. This eliminates challenges such as forgetting to

> free the memory allocated to an object. AMM also avoids the severe danger of using memory

Reviewer: ```suggestion
free the memory allocated to an object. AMM also avoids the critical issue of using memory
```

that is already released. Automatic memory management takes many forms, one of them being
garbage collection (GC).

The garbage collector usually has two responsibilities;

#. Allocate new objects in available memory.
#. Free unused memory.

There are many GC algorithms but MicroPython uses the 
`Mark and Sweep <https://www.geeksforgeeks.org/mark-and-sweep-garbage-collection-algorithm/>`_
policy for managing memory. This algorithm has a mark phase that traverses the heap marking all
live objects while the sweep phase goes through the heap reclaiming all unmarked objects.

.. note::
   The garbage collector automatically runs on the Linux port but may need to be manually 

>    enabled for other ports.

Reviewer: I thought it is enabled for all ports?

Garbage collection functionality in MicroPython is available through the ``gc`` built-in
module:

.. code-block:: console
   
   >>> x = 5
   >>> x
   5
   >>> import gc
   >>> gc.enable()
   >>> gc.mem_alloc()
   1312
   >>> gc.mem_free()
   2071392
   >>> gc.collect()
   19
   >>> gc.disable()
   >>> 

Even when ``gc.disable()`` is invoked, collection can be triggered with ``gc.collect()``.

The object model
~~~~~~~~~~~~~~~~

The structure of a MicroPython object is such that is takes up a word-size. That is to say, 

> pointers and addresses take up a machine word of 8 bytes. Pointers can be 8, 16 to 64 bits.

Reviewer: ```suggestion
pointers and addresses take up a machine word of 8 bytes. Pointers can be 8, 16 or 64 bits.
```

``********|********|********|********|********|********|********|********<tag><tag><tag>``

In a 1 byte address, the 61 bits will hold a value while the 3 lower bits will hold a tag.
This brings us to a concept that MicroPython supports that involves applying a tag to a pointer.

**Pointer Tagging**

> Often than not the lower 3 bits of a pointer are zeros i.e:

Reviewer: ```suggestion
More often than not the lower 3 bits of a pointer are zeroes i.e:
```

Or, maybe, "Frequently, the lower 3 bits are zeroes".

``********|********|********|********|********|********|********|********000``

These bits are reserved for purposes of storing a tag. A tag is a place holder that is used
to store extra information as opposed to introducing a new field to store information usually 
in the object which may be inefficient. 

> The tags can store information like if we are dealing with a small integer, interned(small)

Reviewer: ```suggestion
The tags can store information like if we are dealing with a small integer, interned (small)
```

string or a concrete object as different semantics apply to each of these.

For small integers the mapping is this:

``********|..|******01``

For a small or interned string:

``********|..|******10``

While a concrete object that is neither a small integer nor an interned string takes this form:

``********|..|******00``

Allocation of objects
~~~~~~~~~~~~~~~~~~~~~~

Small integers take up 8 bytes and will be allocated on the stack and not the heap. This implies

> that the allocation of such integers does not affect the heap. Similarly interned strings are small usually
> less of a length less than 10 are stored as an array.

Reviewer: ```suggestion
that the allocation of such integers does not affect the heap. Similarly, interned strings are small - usually
less of a length less than 10 are stored as an array.
```
